jax/interpreters/pxla.py

```diff
@@ -560,20 +560,25 @@ def _sharded_device_array_constant_handler(c, val, canonicalize_types=True):
 
 ### the xla_pmap primitive and its rules are comparable to xla_call in xla.py
 
+def _arg_spec(val, donate):
+  aval = xla.abstractify(val)
+  return aval, donate
+
 def xla_pmap_impl(fun: lu.WrappedFun, *args, backend, axis_name, axis_size, global_axis_size,
-                  devices, name, mapped_invars):
-  abstract_args = map(xla.abstractify, args)
+                  devices, name, mapped_invars, donate):
+  donate += (False,) * (len(args) - len(donate))  # AD changes args not donate.
```

Comment:
This code feels dangerous, we assume that whoever added args, did so only at the end. Wouldn't it be easier to update donate whereever args are added, so that by the time we get here we know that donate corresponds to args. I think that the logic should be very similar to updating mapped_invars (which was tricky).

Reply:
Ah thanks for noticing this, we should follow the logic in xla.py for donate and raise an error `if any(donate)`. AFAIk this happens only for `jvp(jit(.., donate_argnums=..))`.

Reply:
Yes, `JVPTrace.process_call` adds arguments to an `xla_call` (and `JVPTrace.process_map` adds arguments to an `xla_pmap`), so we should update the `params` we pass to `call_primitive.bind` (and `map_primitive.bind`, respectively) in those functions. However, we should also keep in mind that `JaxprTrace.process_call` (and `JaxprTrace.process_map`) produces **two** calls, one with additional outputs and the other with additional inputs; those arises in grad-of-jit situations, for example.

In particular, in `JVPTrace.process_call` we should extend [`donate` in these `params`](https://github.com/google/jax/blob/0ea22b7e191a8cfabefeae18834ced2eec80670b/jax/interpreters/ad.py#L344) by the length of `nonzero_tangents`. (It would be safe to always mark them as False, though there could be another policy, like the tangents corresponding to donated primals are always themselves donated.) Something unfortunate is that we have several call primitives and that same `process_call` is used for all of them, yet only `xla_call` has the `donate` parameter. As a result, we won't always have a `"donate"` in `params`. I'm sure we can figure out good ways to organize things, though I'm fine with anything (e.g. just checking if `"donate"` is in `params` and updating it if it's there).

In `JVPTrace.process_map`, we should also extend [`donate` in these `new_params`](https://github.com/google/jax/blob/0ea22b7e191a8cfabefeae18834ced2eec80670b/jax/interpreters/ad.py#L373). There `mapped_invars` / `new_mapped_invars` provides a good example of the kind of thing we want to do.

I'll write a follow-up comment about `JaxprTrace.process_call` and `JaxprTrace.process_map`.

Reply:
Have fixed this properly with @mattjj's help, we now correctly handle this following mapped_invars (extending to call as well as map).